libraries/AC_WPNav/AC_WPNav.cpp

```diff
@@ -446,13 +442,15 @@ bool AC_WPNav::advance_wp_target_along_track(float dt)
         const float track_velocity = _inav.get_velocity().dot(track_direction);
         // set time scaler to be consistent with the achievable aircraft speed with a 5% buffer for short term variation.
         track_scaler_dt = constrain_float(0.05f + (track_velocity - _pos_control.get_pos_xy_p().kP() * track_error) / curr_target_vel.length(), 0.1f, 1.0f);
-        // set time scaler to not exceed the maximum vertical velocity during terrain following.
-        if (is_positive(curr_target_vel.z)) {
-            track_scaler_dt = MIN(track_scaler_dt, fabsf(_wp_speed_up_cms / curr_target_vel.z));
-        } else if (is_negative(curr_target_vel.z)) {
-            track_scaler_dt = MIN(track_scaler_dt, fabsf(_wp_speed_down_cms / curr_target_vel.z));
-        }
     }
+
+    update_vel_accel(_terain_vel, _terain_accel, dt, false);
+    shape_vel_accel( _wp_speed_cms * offset_z_scaler, 0.0,
+        _terain_vel, _terain_accel,
+        0.0, _wp_speed_cms,
+        -_wp_accel_cmss, _wp_accel_cmss, _pos_control.get_shaping_tc_xy_s(), dt);
+    float vel_time_scalar = _terain_vel / _wp_speed_cms;
```

Comment:
This introduces the chance of a divide-by-zero so I guess _wp_speed_cm should be replaced here by _wp_desired_speed_xy_cms but then we will also need to change wp_and_spline_init() so that when _wp_desired_speed_xy_cms is initialised it is never set to something below WPNAV_WP_SPEED_MIN